sysvabi64/sysvabi64.rst

```diff
@@ -0,0 +1,1076 @@
+
+.. note::
+
+   Relocations are defined in AAELF64_.
+
+Memory Models
+-------------
+
+The AArch64 A64 instruction set has a number of features and
+constraints which make it desirable to use different code models for
+different sizes of executable or dynamic shared object, to improve
+performance and reduce static code size. The relevant constraints are:
+
+* The literal load and ADR instructions generate a PC-relative address
+  with a range of +/- 1MiB.
+
+* The ADRP instruction generates a PC-relative, page aligned address
+  with a range of +/- 4GiB (where page = 4KiB).
+
+* The LDR and STR instructions accept a 12-bit unsigned immediate
+  offset, scaled by the access size.
+
+* The LDR and STR instructions accept 64-bit or 32-bit (signed or
+  unsigned) register offset.
+
+* The BL instruction has a range of +/-128MiB. This is typically used
+  for function / procedure calls.
+
+The following code models are defined
+
+.. table::
+
+  +-------+----------------------+------------+
+  | Code  |    Segment Size      | GOT Limits |
+  | Model |                      |            |
+  +=======+======================+============+
+  | tiny  | text+got+data < 1MiB | 1MiB       |
+  +-------+----------------------+------------+
+  | small | text+got+data < 4GiB | 4GiB       |
+  +-------+----------------------+------------+
+  | large | no assumptions       | 4GiB       |
+  +-------+----------------------+------------+
+
+.. note::
+
+  The segment size column describes the total span of the statically
+  allocated text and data, for example (``__end`` -
+  ``__text_start``). It says nothing about the base address of the
+  program or shared object, which may be located anywhere within the
+  AArch64 virtual address space.
+
+  The text segment includes the sharable PLT, code and read-only data
+  sections.
+
+  The data segment contains the statically defined, writeable,
+  per-process data sections. In all models dynamically allocated data
+  and stack can make use of the full virtual address space, dependent
+  on operating system addressing limits.
+
+  While designing these memory models it was estimated that only 2.6%
+  of load modules (executables and dynamic shared objects) have a
+  total static text+got+data size greater than 1MiB; the rest would
+  all fit into the tiny model. However to avoid Makefile changes, it
+  is recommended that the small model should be the default, with an
+  explicit option to select the tiny model.
+
+  Executables and shared objects may be linked dynamically with other
+  shared objects which use a different code model. However to maintain
+  the properties of the code-model static libraries must be compiled
+  using a model at least as large as the code with which they will be
+  linked.
+
+  The convention for command-line option to select code model is
+  ``-mcmodel=<model>`` where code model is one of small, medium or
+  large.
+
+Running Example
```

Comment:
Missing: TLS examples - or will this go into a separate chapter?

Reply:
They will go in a separate chapter. Still TBD as I thought it better not to wait until I can find the time to write them all up. They are at least as big, if not bigger than the code-models.